source/grid/grid_tools.cc

```diff
@@ -2873,6 +2873,255 @@ namespace GridTools
 
 
 
+
+  template <class Container>
+  std::map< types::global_dof_index,std::vector<typename Container::active_cell_iterator> >
+  get_dof_to_support_patch_map(Container& dof_handler)
+  {
+
+    //  TODO: Add Assert( fe is not dg)
+
+    // This is the map from global_dof_index to
+    // a set of cells on patch.  We first map into
+    // a set because it is very likely that we
+    // will attempt to add a cell more than once
+    // to a particular patch and we want to preserve
+    // uniqueness of cell iterators. std::set does this
+    // automatically for us.  Later after it is all
+    // constructed, we will copy to a map of vectors
+    // since that is the prefered output for other
+    // functions.
+    std::map< types::global_dof_index,std::set<typename Container::active_cell_iterator> > dof_to_set_of_cells_map;
+
+    std::vector<types::global_dof_index> local_dof_index;
+    std::vector<types::global_dof_index> local_face_dof_index;
+    std::vector<types::global_dof_index> local_line_dof_index;
+
+    // in 3d, we need pointers from active lines to the
+    // active parent lines, so we construct it as needed.
+    std::map<typename Container::active_line_iterator, typename Container::line_iterator > lines_to_parent_lines_map;
+    if (Container::dimension == 3)
+    {
+      typename Container::active_cell_iterator cell = dof_handler.begin_active(),
+      endc = dof_handler.end();
+      for (; cell!=endc; ++cell)
+      {
+        // We only want lines that are locally_relevant
+        // although it doesn't hurt to have lines that
+        // are children of ghost cells since there are
+        // few and we don't have to use them.
+        if (cell->is_artificial() == false)
+        {
+          for (unsigned int l=0; l<GeometryInfo<Container::dimension>::lines_per_cell; ++l)
+            if (cell->line(l)->has_children())
+              for (unsigned int c=0; c<cell->line(l)->n_children(); ++c)
+              {
+                lines_to_parent_lines_map[cell->line(l)->child(c)] = cell->line(l);
+                // set flags to know that child
+                // line has an active parent.
+                cell->line(l)->child(c)->set_user_flag();
```

Comment:
if you set user flags, you will need to store all line user flags at the beginning of the function and restore them at the end. Look for other functions that call `save_user_flags()` or `save_user_flags_line()` and later `restore_user_flags()`. The point is that you declared the container object to be `const`, so we shouldn't be modifying things inside (or if we do, restore them to their previous values).